### PR TITLE
[Feat 🔥 ] Frame

### PR DESCRIPTION
**Pull request by hanson-hschang**
Add three things:

- add Frame class: able to set the end frame
- add set_view_distance method: zoom in or out
- change version to VERSION: and add type hint Final to show it is a constant

### CODE REVIEW

**Comment by skim0119** [comment]
> change version to VERSION

I don't see why this change is necessary.

Can you give an example of how to use these features in the documentation?

**Comment by hanson-hschang** [comment]
All caps variable is used to define constant?
Or maybe “version” key word has its own style guide?

It can be used to check if the version is what you want.


BSR_AVALIBILITY = True
try:
    import bsr
    if bsr.VERSION <= “0.1.0”:
        raise ImportError
except ImportError:
    BSR_AVAILABLE = False

if BSR_AVALIBITY:
    bsr.clear_mesh_objects()

Something like this.


However, it doesn’t really matter to me wether it is all caps or not. Let’s
use one that is more commonly used. However, the Final type hint should be
added in there, I think.



Heng-Sheng Chang
Teaching Assistant & Graduate Research Assistant
Mechanical Science and Engineering & Coordinated Science Laboratory
University of Illinois at Urbana-Champaign

File: src/bsr/__init__.py [hunk 1]
```diff
@@ -21,4 +25,57 @@ def get_version() -> str:
         return "unknown"
 
 
-version: str = get_version()
+VERSION: Final[str] = get_version()
+
+
+class Frame:
```
Comment (skim0119) [inline src/bsr/__init__.py:31]: Need documentation. Also it can move to frame.py

File: src/bsr/__init__.py [hunk 2]
```diff
@@ -21,4 +25,57 @@ def get_version() -> str:
...
+            assert (
+                isinstance(frame, int) and frame >= 0
+            ), "frame must be a positive integer or 0"
+        bpy.context.scene.frame_end = frame
+
+
+frame = Frame()
+
+
+def find_area(area_type: str) -> Optional[bpy.types.Area]:
```
Comment (skim0119) [inline src/bsr/__init__.py:65]: Lets add documentation, and move `find_area` and `set_view_distance` to `viewport.py`.

**Review by skim0119 — CHANGES_REQUESTED** [review]

**Review by skim0119 — APPROVED** [review]
LGTM

**Comment by skim0119** [comment]
@hanson-hschang Thanks for fixing the issue

**Comment by hanson-hschang** [comment]
@skim0119 Please review the code again~ I change the implementation of singleton from Mixin to metaclass. It is easier to implement the classes that requires singleton property. Don't have to do the following in the `__init__` method
```python
def __init__(self):
    if not self.isInstantiated:
        self.__frame: int = 0
```
Directly, one can use the following instead
```python
def __init__(self):
    self.__frame: int = 0
```

**Comment by skim0119** [comment]
@hanson-hschang Okay, I'll merge after some test

**Review by skim0119 — APPROVED** [review]

**Pull request merged**